common/djangoapps/third_party_auth/models.py

```diff
@@ -288,7 +306,7 @@ class SAMLProviderConfig(ProviderConfig):
     Configuration Entry for a SAML/Shibboleth provider.
     """
     prefix = 'saml'
-    KEY_FIELDS = ('idp_slug', )
+    KEY_FIELDS = ('site_id', 'idp_slug',)
```

Comment:
For `SAMLProviderConfig`, I don't think we actually need to make `site_id` a `KEY_FIELD`. Instead, administrators can simply set a different `idp_slug` for each site (`university1-default`, `university2-othersite`).

Same might apply for LTI.

For OAuth2, it's necessary because OAuth2 providers are currently keyed by backend name (e.g. Google) rather than some sort of custom identifier (so if we didn't key by site, there would be no other way to have, say, Google on multiple sites). However, you could change that by adding a new `provider_slug` field to OAuth2Provider (with a data migration that sets it to the `backend_name` for backwards compatibility). Then, the unique provider_id for OAuth2Providers could be "oa2-[provider_slug]" instead of "oa2-[backend_name]". This would be more flexible and more consistent with SAML.

With such a change, you wouldn't need to make `site_id` a KEY_FIELD for any of the providers, and this PR would get a lot simpler. You wouldn't need to change the shared ConfigurationModel code at all.